db: Update database models for compatibility with 2.0.12 / HEAD

tuxuser opened the pull request:
This change makes hyperion.rs compatible with latest hyperion.ng database, so settings can be imported from `hyperion.db` and dumped into TOML-representation if desired.

tuxuser committed:
DB: Update DB models to match hyperion.ng 2.0.13-beta.1

tuxuser committed:
DB/TOML: First serialize LedConfig->led_blacklist array, then subtables (fixes issue #5)

alixinne commented on src/models/global.rs:
```diff
@@ -70,12 +70,22 @@ impl Default for FramegrabberType {
     #[validate(range(min = 10))]
     pub height: u32,
+    pub fps: u32,
+    pub framerates: String,
```
This looks like it's a list of framerates, maybe it should be a `Vec<u32>`? Do we have a spec/schema for this somewhere?

alixinne commented on src/models/global.rs:
```diff
@@ -70,12 +70,22 @@ impl Default for FramegrabberType {
+    pub fps: u32,
+    pub framerates: String,
+    pub input: u32,
+    pub resolutions: String,
```
Same as for framerates, this looks like a list, maybe it should be a Vec<Resolution> with Resolution suitably defined?

alixinne commented on src/models/global.rs:
```diff
@@ -70,12 +70,22 @@ impl Default for FramegrabberType {
     #[serde(rename = "type")]
     pub ty: FramegrabberType,
+    #[serde(rename = "available_devices")]
+    pub available_devices: String,
```
Same as below, this should be a list if it's a list of devices, not a String

alixinne commented on src/models/global.rs:
```diff
@@ -70,12 +70,22 @@ impl Default for FramegrabberType {
+    pub available_devices: String,
+    pub device: String,
+    #[serde(rename = "device_inputs")]
+    pub device_inputs: String,
```
Same as below, this should be a list if it's a list of devices, not a String

alixinne commented on src/models/global.rs:
```diff
@@ -120,6 +138,8 @@ pub struct General {
     pub name: String,
     pub watched_version_branch: WatchedVersionBranch,
     pub show_opt_help: bool,
+    pub previous_version: String,
```
Maybe use something like `Option<versions::Version>` from the [versions crate](https://docs.rs/versions/latest/versions/) ?

tuxuser replied:
Sounds nice, it looks very forgiving when parsing versions :D

alixinne replied:
Haha true, it might be a bit too strict, I haven't had time to look at the version formatting hyperion.ng uses (although the versions crate seems to support a lot of different formats, including the appropriately-named [mess](https://docs.rs/versions/latest/versions/struct.Mess.html)).

Since we're not going to handle database/settings migrations anytime soon I feel like an Option<String> would be more than enough in this case.

tuxuser replied:
`Option<String>` that is?

alixinne commented on src/models/global.rs:
```diff
@@ -150,20 +172,33 @@ impl Default for V4L2Standard {
     pub standard: V4L2Standard,
     pub width: u32,
     pub height: u32,
+    pub encoding: String,
```
This should be an enum, like it is done for the standard field.

tuxuser replied:
True that:

https://github.com/hyperion-project/hyperion.ng/blob/master/src/hyperion-v4l2/hyperion-v4l2.cpp#L146
   |
   v
https://github.com/hyperion-project/hyperion.ng/blob/master/include/utils/PixelFormat.h#L71

```rs
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "UPPERCASE", deny_unknown_fields)]
pub enum Pixelformat {
    #[serde(rename = "NO_CHANGE")]
    NoChange,
    Yuyv,
    Uyvy,
    Bgr16,
    Bgr24,
    Rgb32,
    Bgr32,
    I420,
    Nv12,
    Mjpeg,
}

impl Default for Pixelformat {
    fn default() -> Self {
        Self::NoChange
    }
}
```

alixinne replied:
Digging through the source to find actual config values is the hard part of this rewrite tbh, good find!

alixinne commented on src/models/global.rs:
```diff
@@ -150,20 +172,33 @@ impl Default for V4L2Standard {
     pub width: u32,
     pub height: u32,
+    pub encoding: String,
+    pub flip: String,
```
This should be an enum, like it is done for the standard field.

tuxuser replied:
Yep, like above: https://github.com/hyperion-project/hyperion.ng/blob/master/include/utils/PixelFormat.h#L151

```rs
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase", deny_unknown_fields)]
pub enum FlipMode {
    #[serde(rename = "NO_CHANGE")]
    NoChange,
    Horizontal,
    Vertical,
    Both,
}
```

alixinne commented on src/models/instance.rs:
```diff
@@ -295,9 +295,11 @@ impl Default for ForegroundEffect {
 #[serde(default, rename_all = "camelCase", deny_unknown_fields)]
 pub struct InstanceCapture {
     pub system_enable: bool,
+    pub system_grabber_device: String,
```
Looking at the original source, it seems they're storing "NONE" in the config to signify there is no value. I think we could keep this saner by using a `Option<String>` with a `#[serde(deserialize_with = "...")]` attribute?

tuxuser replied:
Would something like that be alright?

```rs
// src/serde/config_option.rs
use std::fmt;

/// Serde visitor for deserializing string values which might
/// be literally a "NONE"-string
struct ConfigOptionalVisitor;

impl<'a> serde::de::Visitor<'a> for ConfigOptionalVisitor {
    type Value = Option<String>;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("A useful type-string or \"NONE\"-string")
    }

    fn visit_str<A>(self, string: &str) -> Result<Self::Value, A>
    where
        A: serde::de::Error,
    {
        match string.to_lowercase() {
            "none" => None,
            _ => Some(string.to_owned()),
        }
    }
}

/// Decode a ConfigOptional value
///
/// # Parameters
///
/// `deserializer`: Serde deserializer
pub fn from_config_optional<'de, D>(
    deserializer: D,
) -> std::result::Result<Option<String>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    deserializer.deserialize_str(ConfigOptionalVisitor {})
}
```

tuxuser replied:
Commited corrected version of `ConfigOptionVisitor`

alixinne commented on src/models/instance.rs:
```diff
@@ -295,9 +295,11 @@ impl Default for ForegroundEffect {
     #[validate(range(min = 100, max = 253))]
     pub system_priority: i32,
     pub v4l_enable: bool,
+    pub v4l_grabber_device: String,
```
Same as for system_grabber_device

tuxuser replied:
Done

tuxuser commented:
When you are suggesting changes from `String` to `Vec<..>`, it would need a custom De-/serializer, correct?
Looking through the schemas in hyperion.ng they list the mentioned fields as  "String".

That are:
- framegrabber->available_devices
- framegrabber->device_inputs
- framegrabber->framerates
- framegrabber->resolutions

Here's the schemas btw: https://github.com/hyperion-project/hyperion.ng/tree/master/libsrc/hyperion/schema

tuxuser committed:
DB: Implement FlipMode and Pixelformat enums

alixinne commented on src/models/global.rs:
```diff
@@ -147,23 +169,74 @@ impl Default for V4L2Standard {
+
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "UPPERCASE", deny_unknown_fields)]
+pub enum Pixelformat {
```
Should be `PixelFormat` with an uppercase `F`, to stay consistent

tuxuser replied:
Ok sure

tuxuser committed:
Corrected capitalization on Pixel(F)ormat enum

tuxuser committed:
DB: Define General->previous_version as optional String type

vtavernier commented:
> When you are suggesting changes from `String` to `Vec<..>`, it would need a custom De-/serializer, correct? Looking through the schemas in hyperion.ng they list the mentioned fields as "String".
> 
> That are:
> 
>     * framegrabber->available_devices
> 
>     * framegrabber->device_inputs
> 
>     * framegrabber->framerates
> 
>     * framegrabber->resolutions
> 
> 
> Here's the schemas btw: https://github.com/hyperion-project/hyperion.ng/tree/master/libsrc/hyperion/schema

Yes, I think we need (at least) a custom deserializer, maybe a serializer if we intend to write back to the database. I had a look at the schema and it does say string, but parts of the JS code for the UI seem to treat this as some kind of list? e.g. 
https://github.com/hyperion-project/hyperion.ng/blob/2f573a117feb34d2c0de6143b0c49466622b45f8/assets/webconfig/js/content_grabber.js#L288

To give more context on the config loading, I went with the "make the parsing as strict as possible", so we only have to do the hard reverse engineering work once, and then the rest of the engine has ready-to-use explicit values (hence the enums, Options, etc. which are not present in the original schema).

tuxuser committed:
feat: Implement ConfigOptionalVisitor serde deserializer

Deserializes string value into Option<String>
-> `None` if string is "None" (case-insensitive check) else `Some(str)`

tuxuser commented:
Did not have the time yet to look into those pseudo-list config options ..

vtavernier commented:
So here's what I gathered from a fresh 2.0.12 install:
 ```json
 {
  "available_devices": "UVC Camera (046d:0825)",
  "blueSignalThreshold": 0,
  "cecDetection": false,
  "cropBottom": 0,
  "cropLeft": 0,
  "cropRight": 0,
  "cropTop": 0,
  "device": "/dev/video1",
  "device_inputs": "0",
  "enable": true,
  "encoding": "YUYV",
  "flip": "NO_CHANGE",
  "fps": 15,
  "fpsSoftwareDecimation": 0,
  "framerates": "15",
  "greenSignalThreshold": 100,
  "hardware_brightness": 128,
  "hardware_contrast": 32,
  "hardware_hue": 0,
  "hardware_saturation": 32,
  "height": 480,
  "input": 0,
  "noSignalCounterThreshold": 200,
  "redSignalThreshold": 0,
  "resolutions": "8",
  "sDHOffsetMax": 0.46,
  "sDHOffsetMin": 0.4,
  "sDVOffsetMax": 0.9,
  "sDVOffsetMin": 0.1,
  "signalDetection": false,
  "sizeDecimation": 8,
  "standard": "NONE",
  "width": 640
}
```
It seems that even though the field names are plural, they only represent a single value in the UI (there's only one of each capture device type selected for any single instance), so they're not lists, and their type is as follows:
* `available_devices`: `String`, friendly device name
* `device_inputs`: `usize` (but stored as a String, needs parsing), index into the list of known V4L2 devices?
* `framerates`: `u32` (but stored as a String, needs parsing), number of frames per second, matches what's in the `fps` field
* `resolutions`: `usize` (but stored as a String, needs parsing), index into the list of resolutions supported by the V4L2 device. The actual resolution is given by `width` and `height`

tuxuser commented:
Thx for posting the json representation of the config section in question.

For the deserialization `FromStr` and serialization `fmt::Display` an additional dependency could be pulled in: https://crates.io/crates/serde_with

```rs
#[serde_as]
#[derive(Deserialize, Serialize)]
struct Foo {
    // Serialize with Display, deserialize with FromStr
    #[serde_as(as = "DisplayFromStr")]
    bar: u8,
}
```

vtavernier commented:
> Thx for posting the json representation of the config section in question.
> 
> For the deserialization `FromStr` and serialization `fmt::Display` an additional dependency could be pulled in: https://crates.io/crates/serde_with
> 
> ```rust
> #[serde_as]
> #[derive(Deserialize, Serialize)]
> struct Foo {
>     // Serialize with Display, deserialize with FromStr
>     #[serde_as(as = "DisplayFromStr")]
>     bar: u8,
> }
> ```

Good idea, I wasn't aware of `serde_with`, it makes sense for these.

tuxuser committed:
DB: De-/Serialize certain numeric values in models::Framegrabber as string

tuxuser commented:
~~`Framegrabber` seems like the base struct and `GrabberV4L2` derives from it.
The json you posted has fields from `GrabberV4L2`.

Now, to keep code duplication at a minimum, shall we pull in `Framebuffer` into `GrabberV4L2` via serde-flatten, like so:

```rs
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Validate)]
#[serde(default, rename_all = "camelCase", deny_unknown_fields)]
pub struct GrabberV4L2 {
   ...
   #[serde(flatten)]
   base: Framegrabber,
}
```~~

Incorrect, V4L2 does not inherit all the fields Framebuffer... annoying!
So it needs fields re-defined...
